SU2_CFD/src/solver_direct_mean.cpp

```diff
@@ -12483,13 +12483,241 @@ void CEulerSolver::BC_Engine_Exhaust(CGeometry *geometry, CSolver **solver_conta
+
+        /*--- Reflect the gradients for all scalars including the velocity components.
+              The gradients of the velocity components are overriden later with the 
+              correct values: grad(V)_r = grad(V) - 2 [grad(V)*n]n, V beeing any primitive ---*/
+        for (iVar = 0; iVar < nPrimVarGrad; iVar++) {
+
+          /*--- Compute projected part of the gradient in a dot product ---*/
+          ProjGradient = 0.0;
+          for (iDim = 0; iDim < nDim; iDim++)
+            ProjGradient += Grad_Prim[iVar][iDim]*UnitNormal[iDim];
+
+          for (iDim = 0; iDim < nDim; iDim++)
+            Grad_Reflected[iVar][iDim] = Grad_Prim[iVar][iDim] - 2.0 * ProjGradient*UnitNormal[iDim];
+        }
+        
+        /*--- Compute unit tangential, the direction is arbitrary as long as t*n=0 ---*/
+        switch( nDim ) {
+          case 2: {
+            Tangential[0] = -UnitNormal[1];
+            Tangential[1] =  UnitNormal[0];
+            break;
+          }
+          case 3: {
+            /*--- Find the largest entry index of the UnitNormal, and create Tangential vector based on that. ---*/
+            unsigned short Largest, Arbitrary, Zero;
+            if     (abs(UnitNormal[0]) >= abs(UnitNormal[1]) && abs(UnitNormal[0]) >= abs(UnitNormal[2])){Largest=0;Arbitrary=1;Zero=2;}
+            else if(abs(UnitNormal[1]) >= abs(UnitNormal[0]) && abs(UnitNormal[1]) >= abs(UnitNormal[2])){Largest=1;Arbitrary=0;Zero=2;}
+            else                                                                                         {Largest=2;Arbitrary=1;Zero=0;}
+
+            Tangential[Largest] = -UnitNormal[Arbitrary]/sqrt(pow(UnitNormal[Largest],2) + pow(UnitNormal[Arbitrary],2));
+            Tangential[Arbitrary] =  UnitNormal[Largest]/sqrt(pow(UnitNormal[Largest],2) + pow(UnitNormal[Arbitrary],2));
+            Tangential[Zero] =  0.0;
+            break;
+          }
+        }
+        
+        /*--- Compute gradients of normal and tangential velocity:
+              grad(v*n) = grad(v_x) n_x + grad(v_y) n_y (+ grad(v_z) n_z)
+              grad(v*t) = grad(v_x) t_x + grad(v_y) t_y (+ grad(v_z) t_z) ---*/
+        su2double GradNormVel[nDim];
+        su2double GradTangVel[nDim];
+        for (iVar = 0; iVar < nDim; iVar++) { // counts gradient components
+          GradNormVel[iVar] = 0.0;
+          GradTangVel[iVar] = 0.0;
+          for (iDim = 0; iDim < nDim; iDim++) { // counts sum with unit normaal
+            GradNormVel[iVar] += Grad_Prim[iDim+1][iVar] * UnitNormal[iDim];
+            GradTangVel[iVar] += Grad_Prim[iDim+1][iVar] * Tangential[iDim];
+          }
+        }
+
+        /*--- Refelect gradients in tangential and normal direction by substracting the normal/tangential
+              component twice, just as done with velocity above.
+              grad(v*n)_r = grad(v*n) - 2 {grad([v*n])*t}t
+              grad(v*t)_r = grad(v*t) - 2 {grad([v*t])*n}n ---*/
+        su2double ReflGradNormVel[nDim];
```

Comment:
The routine is fairly memory intensive. For speed, can some of the allocations be moved out of the loop over all vertices, be made dynamically, or even use some kind of persistent memory from the class data that does not need to be allocated/reallocated so often? Or maybe to cut down on the total required memory, some of the arrays can be reused during different parts of the routine.

Reply:
Moved all allocations out the loop and got rid of some (3) arrays. The readability of the code together with the comments is still given I think